quinn-udp/src/cmsg/mod.rs

```diff
@@ -124,6 +124,10 @@ pub(crate) trait MsgHdr {
 
     fn cmsg_nxt_hdr(&self, cmsg: &Self::ControlMessage) -> *mut Self::ControlMessage;
 
+    /// Sets the number of control messages added to this `struct msghdr`.
+    ///
+    /// Note that this is a destructive operation and should only be done as a finalisation
+    /// step.
```

Comment:
This is perhaps a bit unfortunate.  It works as it is used now but I could not figure out a nice way to actually force this into the type system.

The `struct msghdr` is part of a larger struct here, so adding a finalisation step that takes ownership does not work.

The other option I considered was adding a separate `.finish()` method, but apart from being slightly more explicit it has the same problem with being purely conventional.

Reply:
There's a lot in this module that relies on careful usage, so I think this is reasonable as-is, though of course improvements are always welcome.